File: mbox-parser.py
import argparse
import getopt
import re
import sys


def usage():
    print('cat message.mbox | mbox-parser.py -a aliases.txt')


def parse_mbox_from_stdin(aliases):
    lines = sys.stdin.readlines()
    name_and_email_dict = {}
    not_use_aliases = not aliases
    for line in lines:
        if line.startswith('From:'):
            name_match = re.search('(?<=From: )[a-zA-Z ]+(?!\S)', line)
            email_match = re.search('<(.*?)>', line)
            if email_match:
                name = name_match.group(0) if name_match else ''
                email = email_match.group(1)
                if not_use_aliases:
                    if email in name_and_email_dict:
                        count = name_and_email_dict[email][-1]
                        name_and_email_dict[email] = (name, email, count + 1)
                    else:
                        name_and_email_dict[email] = (name, email, 1)
                else:
                    has_alias = [name == alias.split(' ')[0] or email == alias.split(' ')[0] for alias in aliases]
                    if has_alias[0] or has_alias[1]:
                        count = 0
                        key = ''
                        if has_alias[0] and name in name_and_email_dict:
                            count = name_and_email_dict[name][-1]
                            key = name
                        elif has_alias[1] and email in name_and_email_dict:
                            count = name_and_email_dict[email][-1]
                            key = email
                        elif has_alias[0]:
                            name_and_email_dict[name] = (name, email, 1)
                            continue
                        elif has_alias[1]:
                            name_and_email_dict[email] = (name, email, 1)
                            continue
                        name_and_email_dict[key] = (name, email, count + 1)
                    elif email in name_and_email_dict:
                        count = name_and_email_dict[email][-1]
                        name_and_email_dict[email] = (name, email, count + 1)
                    else:
                        name_and_email_dict[email] = (name, email, 1)
            else:
                print("Missed email")
                sys.exit(1)

    name_and_email_dict_sorted = sorted(name_and_email_dict.items(), key=lambda e: e[1][2])

    for (k, v) in name_and_email_dict_sorted:
        print(k, v[-1])


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument('-a', '--aliases', help='Псевдонимы пользователей')
    args = parser.parse_args()
    aliases = []
    if args.aliases:
        aliases = [line.rstrip() for line in open(args.aliases)]
    parse_mbox_from_stdin(aliases)

if __name__ == "__main__":
    main()

File: mbox-parser.rb
#! /usr/bin/env ruby

require 'optparse'
require 'ostruct'

class MBoxParser

  def initialize
    @name_and_email_dict = {}
  end

  def parse(messages, aliases_list)
    messages.each do |message|
      if message.start_with? 'From: '
        match = message.match(/From: ([a-zA-Z ]+\s+)?<(.*)>/i)
        if match
          name, email = match.captures
          name = name ? name : ''
          if aliases_list.empty?
            simple_parse name, email
          else
            consider_aliases_parse name, email, aliases_list
          end
        end
      end
    end
  end

  def print_result
    @name_and_email_dict.sort_by { |_, value| value[-1] }.each { |key, value|
      puts key + ': ' + value[-1].to_s
    }
  end

  private
  def simple_parse(name, email)
    if @name_and_email_dict.has_key? email
      @name_and_email_dict[email][-1] += 1
    else
      @name_and_email_dict[email] = name, email, 1
    end
  end

  def consider_aliases_parse(name, email, aliases)
    has_aliases = has_aliases email, aliases
    if has_aliases[0]
      key = has_aliases[1]
      if @name_and_email_dict.has_key? key
        @name_and_email_dict[key][-1] += 1
      else
        @name_and_email_dict[key] = key, email, 1
      end
    elsif @name_and_email_dict.has_key? email
      @name_and_email_dict[email][-1] += 1
    else
      @name_and_email_dict[email] = name, email, 1
    end
  end

  def has_aliases(email, aliases)
    aliases.each { |alias_line|
      alias_list = alias_line.split ' '
      if alias_list.include? email
        return TRUE, alias_list[0]
      end
    }
    return FALSE, email
  end
end

if __FILE__ == $0
  messages = []
  line = $stdin.gets
  while line
    messages << line.strip
    line = $stdin.gets
  end

  options = OpenStruct.new
  options.aliases = ''
  opt_parser = OptionParser.new do |opts|
    opts.banner = 'Запуск: cat messages.mbox | mbox-parser.rb [-a aliases.txt]'
    opts.on('-a', '--aliases ALIASES') do |a|
      options.aliases = a
    end
  end
  opt_parser.parse! ARGV

  parser = MBoxParser.new

  aliases_list = []
  unless options.aliases.empty?
    begin
      aliases_list = File.open(options.aliases) { |file| file.readlines }
    rescue
      puts 'Файл с псевдонимами не найден. Выполняем парсинг без учета псевдонимов.'
    end
  end
  parser.parse messages, aliases_list
  parser.print_result
end
